### ISSUE TITLE
sql: increase coverage of SQL module by `diag_set()`

### ISSUE THREAD

**Issue by LeonidVas** [tarantool/tarantool#5624]
We now have a situation where `sql_execute()`  returns -1, but the `diag` is empty(https://github.com/tarantool/tarantool/issues/5537). This will lead to `SEGFAULT` inside `lbox_execute()` which is not nice.
Seems like we shouldn't have an empty `diag` anyway if `sql_execute()` fails.
It might be cool to use `stack diag` is this module. This will allow to see both the high-level problem that is understandable to the user and its cause.

**Comment by Totktonada** [comment]
I don't understand the goal:

1. Whether it is about spotting incorrect code paths, where `diag_set()` should be used, but it does not?
2. Whether it is about better errors reporting (in particular using the stacked diagnostics)?

**Comment by LeonidVas** [comment]
> I don't understand the goal:
> 
> 1. Whether it is about spotting incorrect code paths, where `diag_set()` should be used, but it does not?

Yep. For example: https://lists.tarantool.org/pipermail/tarantool-patches/2020-December/021406.html

> 2. Whether it is about better errors reporting (in particular using the stacked diagnostics)?

**Comment by LeonidVas** [comment]
> I don't understand the goal:
> 
> 1. Whether it is about spotting incorrect code paths, where `diag_set()` should be used, but it does not?

Yep. For example: https://lists.tarantool.org/pipermail/tarantool-patches/2020-December/021406.html
If  `sql_execute()` failed (return -1), this may be the reason of a crash inside `luaT_pusherror()` (we don't check `e != NULL` in the release build, this is our policy, as I understand).

> 2. Whether it is about better errors reporting (in particular using the stacked diagnostics)?

Yep.  I think it's a more friendly to have a high-level description of the problem for the customer (like: "sql execute is failed") and a  low-level description for maintainers (like:"failed to create a temporary file.")

**Comment by Totktonada** [comment]
So it is two different problems in fact. One re segfault, another re verbosity of errors.

Regarding the first one. How you propose to spot all such places? Just look around the code? At least it should be a certain part of code to make the task measurable.

Regarding the second: the similar problem here. The task is not measurable.

**Comment by LeonidVas** [comment]
"... two different problems in fact" - IMHO, no it is not.so.
I suggest to move from high-level to low level. It just more simple because then you can add `diag_add()` in two or three places on exit and after add `diag_set()` with low-level errors.

>  Just look around the code?

Do you have any better ideas?

> ...to make the task measurable.

I think this is a first part of the task: to deal which part we want to cover with `diag`.

**Comment by Totktonada** [comment]
> Do you have any better ideas?

Manually mark functions as a reporters to the diagnostics area and implement a static analysys, which forbids to leave from a such function with a negative value without setting a diagnostics. Just like checked exceptions in Java. The implementation of such static analysys may be complicated.

AFAIK, you spotted several places and just want to fix them within this task. So, it worth to share your observations and consider the task as fixing of them and maybe near places of this kind.

**Comment by LeonidVas** [comment]
> AFAIK, you spotted several places and just want to fix them within this task. So, it worth to share your observations and consider the task as fixing of them and maybe near places of this kind.

Yep, while fixing the crash, several places was spotted and @sergos  suggested creating a task to continue working in this direction in the future. And I created it)
